// reactjs-components/lib/Dropdown/DropdownListTrigger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require("babel-runtime/core-js/object/get-prototype-of");

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _DropdownTrigger2 = require("./DropdownTrigger");

var _DropdownTrigger3 = _interopRequireDefault(_DropdownTrigger2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DropdownListTrigger = function (_DropdownTrigger) {
  (0, _inherits3.default)(DropdownListTrigger, _DropdownTrigger);

  function DropdownListTrigger() {
    (0, _classCallCheck3.default)(this, DropdownListTrigger);
    return (0, _possibleConstructorReturn3.default)(this, (DropdownListTrigger.__proto__ || (0, _getPrototypeOf2.default)(DropdownListTrigger)).apply(this, arguments));
  }

  (0, _createClass3.default)(DropdownListTrigger, [{
    key: "render",
    value: function render() {
      var _props = this.props,
          className = _props.className,
          disabled = _props.disabled,
          selectedItem = _props.selectedItem;

      var html = selectedItem ? selectedItem.selectedHtml || selectedItem.html : this.props.placeholder || null;

      return _react2.default.createElement(
        "button",
        {
          className: className,
          disabled: disabled,
          onClick: this.handleTrigger,
          type: "button"
        },
        html
      );
    }
  }]);
  return DropdownListTrigger;
}(_DropdownTrigger3.default);

exports.default = DropdownListTrigger;


DropdownListTrigger.defaultProps = {
  className: "",
  placeholder: "",
  selectedItem: null,
  onTrigger: function onTrigger() {},

  disabled: false
};

DropdownListTrigger.propTypes = {
  className: _propTypes2.default.string,
  placeholder: _propTypes2.default.string,
  selectedItem: _propTypes2.default.object,
  onTrigger: _propTypes2.default.func,
  disabled: _propTypes2.default.bool
};
